### PR TITLE
HDDS-7734. Implement symmetric SecretKeys lifescycle management in SCM

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

This is the implementation of symmetric SecretKeys lifecycle management in SCM, with SecretKeys synchronized between SCMs using Ratis (a few notes for SecretKey replication are [here](https://docs.google.com/document/d/125D83cFu47AslupJ-3k0FRskgzpNbz4hzDC00dV5-O4/edit#heading=h.vzi1ddqxwlg0)).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7734

## How was this patch tested?

Unit-tested. Integration test will come per [HDDS-7830)](https://issues.apache.org/jira/browse/HDDS-7830)

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SerializableCodec.java
Comment (fapifta) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SerializableCodec.java:19]: Shouldn't this be in the o.a.h.hdds.scm.ha.io.codec package as the other codec implementations?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SerializableCodec.java
Comment (jojochuang) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SerializableCodec.java:39]: We should stay away from ByteArrayOutputStream and ObjectOutputStream.
(1) Potential vulnerability due to code injection.
(2) No guarantee of compatibility.

**Review by jojochuang — CHANGES_REQUESTED** [review]

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java
Comment (jojochuang) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java:108]: should it retry if time out? RuntimeException would crash the SCM. Probably makes sense for such a critical service.

**Review by fapifta — COMMENTED** [review]
@duongkame thank you for working on this one, the code and the structure of the code looks well thought out.

As we discussed with @jojochuang we should not use Serializable, instead we probably should serialize and deserialize the things with protobuf as we do in other parts of the code.

Besides this please find my comments inline.

**Review by kerneltime — COMMENTED** [review]
Currently, the code is differentiating current keys and all keys and has a lot of code to make the distinction between them. Would keeping the keys to always be a sorted list be cleaner? The index 0 or n-1 is always the current key and we copy and move the entire list around in the code as needed.
Half way done will take a deeper look at ratis integration and tests later.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java
Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java:109]: I like using all functionals too, something like`dto.forEach(writer::write)`, yet, the `write` method throws `IOException`, and the checked exception is not "functional" friendly. A wrapper to turn it into RuntimeException can work, but it looks like an unnecessary complication.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SerializableCodec.java
Comment (duongkame) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SerializableCodec.java]: Make sense, I've updated to use protobuf.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/package-info.java
Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/package-info.java:63]: Totally, I've updated it with more detailed documentation. Would be great if you can have another look.

File: hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStoreTest.java
Comment (duongkame) [reply hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStoreTest.java:84]: Good point, compatible test has been added.

**Comment by duongkame** [comment]
> Currently, the code is differentiating current keys and all keys and has a lot of code to make the distinction between them. Would keeping the keys to always be a sorted list be cleaner? The index 0 or n-1 is always the current key and we copy and move the entire list around in the code as needed. Half way done will take a deeper look at ratis integration and tests later.

Thanks for the suggestion, @kerneltime. It was started in the most natural way which apparently not the smartest one. All the code related to the distinction has been cleaned up.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java
Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java:127]: I believe all the time durations like rotation or expiry should be much longer than the accepted time skew. 

Plus, a SecretKey is always generated on top of the existing agreed ones (we use the latest key timestamp to determine if it's time to generate a new key). And that way, the time order between SecretKey is always protected.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java
Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java:116]: `FileAlreadyExistsException` is a subclass of `IOException`, maybe we don't need to include both?

Comment (kerneltime) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java:116]: Makes sense unless we want to differentiate. What would the error handling be if this step fails? Will go over it again.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/SecretKeyManager.java
Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/SecretKeyManager.java]: Yup. As discussed offline, the keys should be agreed upon by all SCM parties before being visible to clients. A change has been made.

**Comment by duongkame** [comment]
> @duongkame thank you for working on this one, the code and the structure of the code looks well thought out.
> 
> As we discussed with @jojochuang we should not use Serializable, instead we probably should serialize and deserialize the things with protobuf as we do in other parts of the code.
> 
> Besides this please find my comments inline.

Thanks for the review @fapifta. I visited all the comments and took action.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsConfigKeys.java
Comment (duongkame) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsConfigKeys.java:219]: I added an explanation in `ozone-site.xml`

File: hadoop-hdds/common/src/main/resources/ozone-default.xml
Comment (duongkame) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml:3648]: It's updated.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java
Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/LocalSecretKeyStore.java:122]: Any `IOException` will result in SCM crash because the secret keys cannot be saved to files as a reaction to a Ratis log entry.

As we check for the file existence before creating them, the window of error is small (as someone may create the same file at the very same time). Crashing SCM for this unlikely event is just fine to me.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java
Comment (duongkame) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java:106]: A `TimeoutException` translated to `RuntimeException` would crash the scheduled task, not SCM. 
And a `TimeoutException` doesn't indicate a failure, it just means the wait is over, and the ratis update flow would be still happening asynchronously. 

The next rotation check will check if the state is initialized by ratis, otherwise, it'll retry. 
Thay may, however, leave a window in which SecretKey state may not have been initialized and tokens can't be generated.  For now, the rotation check is configured to run every 10m.

**Comment by hemantk-12** [comment]
@jojochuang and @fapifta Can you please look at the latest revision? Thanks.

File: hadoop-hdds/common/src/main/resources/ozone-default.xml
Comment (fapifta) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:3648]: I think this line should not be here, the algorithm should be named based on KeyGenerator's accepted names documented here: https://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#KeyGenerator

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java
Comment (fapifta) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java:87]: Should this be part of the SecretKeyConfig?
Also, I think, as we expect a duration in the config, and we work with a duration here in the code, why don't we just read the config as a string and then construct a duration from that instead of converting it to a long than back to a duration?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java
Comment (fapifta) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java:142]: I am not sure if this one should be on info level, I think we will log this every 10 minutes throughout the day by default. Does that hold any meaningful information in the log file?
Also there is one log message from secretKeyManager once the new key is generated, so we might not need the logging here at all, what do you think?

**Review by fapifta — COMMENTED** [review]
Hi @duongkame thank you for the continued work on this one, in general it is getting better and I like the approach and most of the code, though I have added some minor comments inline, please take a look.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/package-info.java
Comment (duongkame) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/symmetric/package-info.java]: Good point. I pinned the link to the jira containing the design doc.

File: hadoop-hdds/common/src/main/resources/ozone-default.xml
Comment (duongkame) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml]: My bad. Updated the description.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java
Comment (duongkame) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java]: Originally I wanted to keep `SecretKeyConfig` enclosing configs used by the `org.apache.hadoop.hdds.security.symmetric` only. Yet, eventually it is just a config parser for secret key related logic. Moved.

Regarding reading the duration as a String and converting to `Duration` directly, I think it's a matter of format. `Duration` supports ISO-8601 format, i.e the configured string has to be like `PT1D` or `P1H`... While the standard Hadoop time duration reader supports a custom "in-house" format like `1d`, `2h`... 
Most of the time duration configs (except some configs related to X509 cerificate) are in the Hadoop format and I think we'd prefer to keep it consistent that way.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java
Comment (duongkame) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java]: Even though the logs don't really indicate anything meaningful, I wanted to keep them as the thread "heartbeat".
Yet, I guess you're right. I removed them, and will add a few metrics instead to help monitor SecretKeyService (HDDS-8004).

**Comment by duongkame** [comment]
> 



> Hi @duongkame thank you for the continued work on this one, in general it is getting better and I like the approach and most of the code, though I have added some minor comments inline, please take a look.

Thanks again for the very thorough review, @fapifta. I've looked into them and made the necessary changes.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java
Comment (fapifta) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java]: Ah I see, I haven't think about that... alright, let's leave it this way.

**Comment by fapifta** [comment]
You're welcome!
In the meantime one thought does bother me a bit, and would like to understand the intent for that...

In the SecretKeyManager and the SecretKeyState, I have some doubts around the mechanism that generates a new key and does the rotation... it is hard to tell what exactly, it is just a bad feeling so far, so let me think out loud here:
If I understand correctly, SecretKeyManager will decide it should rotate, once the current key is older than 1 day.
Once the rotation happens eventually after 1d, the SecretKeyManager#checkAndRotate() method creates a new list of secret keys, it filters the oldest expired key from the list, and SecretKeyState#updateKeys is called with the keys in a sorted order, where the newest key is the first item in the list.
This call to updateKeys is replicated to the other SCMs, so all SCMs eventually will have the new set of keys.

So far this is clean on its own, the confusion on my side comes after this. The way of implementation suggests that the second proposal the pull model will be used on the DNs.
How this will look like?
I guess if the DN runs into a token, that is newer then the newest key it has, it will pull the list of the keys from SCM. While OMs will pull periodically from SCM.
If my understanding is correct, then operations wise we are good, and probably my mind stuck with the push model and the pre-generated and distributed next key.



During the thought experiment around this, I started to feel that it might not be correct to have a SecretKeyState that gets its internal representation from outside, and also misses proper encapsulation... With the current code this might not be a problem, but here are the concerns:
1. the internal list is guarded by locking, but its reference is leaking via the getSortedKeys() method.
2. updateKeys can override the full internal state

1. is easy to fix, and probably it is just some automatism where usually I don't think about this concern either for the first sight
2. on the other hand there might be a more interesting question... Let's do some thought experiment:
Let's say we have one SCM (SCM1) that is stopped for an extended period of time, and its raft logs are so old that when it gets back in service it downloads a new snapshot, and since that snapshot there was no rotation happening. At this time SCM2 is the leader, but after SCM1 comes up and syncronizes its data and becomes a proper follower, SCM2 goes down before there was a rotation of secret keys, and during election SCM1 becomes the elected leader.

In this scenario will SCM1 have the secretKeys properly? As based on the code my assumption is that in this case SCM1 will have 1 secret key in its list generated at startup (let's assume every key that was persisted has expired already), with that it will issue tokens, and DNs will be able to validate those tokens, but once keys are rotated, the list of old keys will be overridden in all SCMs.

After this has happened, let's assume a DN goes down, and comes up due to a restart for any reason. In this case if there is a client that has a token issued by SCM2 back when it was the leader, was not be able to read data from the restarted DN, as that DN will not be able to fetch the formerly valid token from anywhere, while those DNs that have the token in memory will work fine further causing intermittent read failures.

Do I miss something, or this is a legit scenario with this approach of managing the secret keys?
I see this happening because the keys are managed in a file, not in rocksDB, and so when SCM1 downloads the snapshot of the data to replay logs from that snapshot, it might not replay the transaction that updates the file, and the file is not in the snapshot, but putting the secret keys to rocksDB with that to the snapshot will hurt us later when it comes to conforming with FISMA  for example.

**Comment by duongkame** [comment]
> You're welcome! In the meantime one thought does bother me a bit, and would like to understand the intent for that...
> 
> In the SecretKeyManager and the SecretKeyState, I have some doubts around the mechanism that generates a new key and does the rotation... it is hard to tell what exactly, it is just a bad feeling so far, so let me think out loud here: If I understand correctly, SecretKeyManager will decide it should rotate, once the current key is older than 1 day. Once the rotation happens eventually after 1d, the SecretKeyManager#checkAndRotate() method creates a new list of secret keys, it filters the oldest expired key from the list, and SecretKeyState#updateKeys is called with the keys in a sorted order, where the newest key is the first item in the list. This call to updateKeys is replicated to the other SCMs, so all SCMs eventually will have the new set of keys.
> 
> So far this is clean on its own, the confusion on my side comes after this. The way of implementation suggests that the second proposal the pull model will be used on the DNs. How this will look like? I guess if the DN runs into a token, that is newer then the newest key it has, it will pull the list of the keys from SCM. While OMs will pull periodically from SCM. If my understanding is correct, then operations wise we are good, and probably my mind stuck with the push model and the pre-generated and distributed next key.
> 
> During the thought experiment around this, I started to feel that it might not be correct to have a SecretKeyState that gets its internal representation from outside, and also misses proper encapsulation... With the current code this might not be a problem, but here are the concerns:
> 
> 1. the internal list is guarded by locking, but its reference is leaking via the getSortedKeys() method. Though the actual implementation used is an unmodifiable list as of now.
> 2. updateKeys can override the full internal state
> 
> The first one might not be a problem at all at the end of the day, it might be, if someone carelessly change the internal representation from unmodifiable list to a mutable list impementation. Second one on the other hand there might be a more interesting question... Let's do some thought experiment: Let's say we have one SCM (SCM1) that is stopped for an extended period of time, and its raft logs are so old that when it gets back in service it downloads a new snapshot, and since that snapshot there was no rotation happening. At this time SCM2 is the leader, but after SCM1 comes up and syncronizes its data and becomes a proper follower, SCM2 goes down before there was a rotation of secret keys, and during election SCM1 becomes the elected leader.
> 
> In this scenario will SCM1 have the secretKeys properly? As based on the code my assumption is that in this case SCM1 will have 1 secret key in its list generated at startup (let's assume every key that was persisted has expired already), with that it will issue tokens, and DNs will be able to validate those tokens, but once keys are rotated, the list of old keys will be overridden in all SCMs.
> 
> After this has happened, let's assume a DN goes down, and comes up due to a restart for any reason. In this case if there is a client that has a token issued by SCM2 back when it was the leader, was not be able to read data from the restarted DN, as that DN will not be able to fetch the formerly valid token from anywhere, while those DNs that have the token in memory will work fine further causing intermittent read failures.
> 
> Do I miss something, or this is a legit scenario with this approach of managing the secret keys? I see this happening because the keys are managed in a file, not in rocksDB, and so when SCM1 downloads the snapshot of the data to replay logs from that snapshot, it might not replay the transaction that updates the file, and the file is not in the snapshot, but putting the secret keys to rocksDB with that to the snapshot will hurt us later when it comes to conforming with FISMA for example.

Thanks again for being thorough, @fapifta. 

#1 is just a design choice. I was considering between a) creating a copy list to serve a query, or b) leveraging immutability. I think in Ozone we're so easy to go with a) and that's something I notice when doing a memory scaling analysis in SCM. Today, buffers, e.g. pipeline datanodes, are just copied over and that seems unnecessary. Immutability is a good alternative to deal with concurrency and should be used where possible. (Yes, I know, it's not that easy to spread that awareness in Java, i.e. there's no good implicit type for immutable data, which would be easier in the functional world).

**Comment by fapifta** [comment]
Hi @duongkame,

Thank you for your answers, I haven't noticed HDDS-7945, and that I believe is entitled to solve the consistency problem I tried to describe. The problem is when SCM2 takes on leadership, and does not get the previous state because it was not part of the snapshot, and the updates are not there during log replay, so after that it would overwrite the keys used with the next rotation. This problem should be solved, if the keys are syncronized within the snapshot as well in between the SCMs.

The first one with the unmodifiable list is perfectly fine this way, I just left the thought there, but did not wanted to propose any change just wanted to discuss and understand how you think about it, thank you for sharing the thought process of yours.

From my side, I think I am fine with the changes. +1.
The only thing remained, that is not to be fixed within the PR itself, but we should get rid of the internal JIRA link in the design doc added to the main JIRA ;)

**Comment by duongkame** [comment]
Thanks, @fapifta. Totally agree that anything replicated by Ratis should be a part of Ratis snapshot, including SecretKeys. 
Also, I'll clean up the design doc and create a final immutable version (pdf).
Just one more thing, can you indeed approve the PR?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java
Comment (kerneltime) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/SecretKeyManagerService.java:98]: We should look into uncaught exception handling. Cashing SCM in the handler makes sense if keys cannot be managed.

**Review by kerneltime — APPROVED** [review]
I have questions around failover testing but that can be addressed in a subsequent PR

**Comment by kerneltime** [comment]
We can merge this in once @fapifta has approved it.
Since we have a developer branch now, we can merge more eagerly and file follow up jiras for work identified.

**Review by fapifta — APPROVED** [review]
@duongkame  thank you for addressing my suggestions. +1

**Comment by duongkame** [comment]
Thanks @fapifta @kerneltime @jojochuang for the review.

**Pull request merged**
